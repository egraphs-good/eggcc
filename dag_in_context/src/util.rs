use std::{
    ffi::OsStr,
    fs::{self, File},
    process::{Command, Stdio},
    time::{SystemTime, UNIX_EPOCH},
};

/// Invokes some program with the given arguments, piping the given input to the program.
/// Returns an error if the program returns a non-zero exit code.
/// Code adapted from https://github.com/egraphs-good/egg/blob/e7845c5ae34267256b544c8e6b5bc36d91d096d2/src/dot.rs#L127
pub fn run_cmd_line<S1, S2, I>(program: S1, args: I, input: &str) -> std::io::Result<String>
where
    S1: AsRef<OsStr>,
    S2: AsRef<OsStr>,
    I: IntoIterator<Item = S2>,
{
    // Write the input to a temporary file so the child can read it directly.
    let mut temp_path = std::env::temp_dir();
    let unique = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos();
    temp_path.push(format!(
        "eggcc-run-cmd-{}-{}.tmp",
        std::process::id(),
        unique
    ));

    fs::write(&temp_path, input)?;
    let input_file = File::open(&temp_path)?;

    let output = Command::new(program)
        .args(args)
        .stdin(Stdio::from(input_file))
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output();

    let output = match output {
        Ok(out) => {
            let _ = fs::remove_file(&temp_path);
            out
        }
        Err(err) => {
            let _ = fs::remove_file(&temp_path);
            return Err(err);
        }
    };

    match output.status.code() {
        Some(0) => Ok(String::from_utf8(output.stdout)
            .map_err(|e| std::io::Error::other(format!("utf8 error: {}", e)))?),
        Some(exit) => {
            let stderr = String::from_utf8_lossy(&output.stderr);
            Err(std::io::Error::other(format!(
                "program returned error code {exit}; stderr: {stderr}"
            )))
        }
        None => {
            let stderr = String::from_utf8_lossy(&output.stderr);
            Err(std::io::Error::other(format!(
                "program was terminated by signal; stderr: {stderr}"
            )))
        }
    }
}
