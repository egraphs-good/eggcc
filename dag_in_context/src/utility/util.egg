(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((Switch pred inputs branch)) ((union (ListExpr-suffix branch 0) branch)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
   (Cons a (Append b e))
   :ruleset always-run)
(rewrite (Append (Nil) e)
   (Cons e (Nil))
   :ruleset always-run)

(function tuple-length (Expr) i64 :unextractable)

(rule ((HasType expr (TupleT tl))
       (= len (TypeList-length tl)))
      ((set (tuple-length expr) len)) :ruleset always-run)

;; Create a Get for every index, and rewrite it to see through Concat
(rule ((Single expr)) ((union (Get (Single expr) 0) expr)) :ruleset always-run)
;; initial get
(rule ((Concat expr1 expr2)
       (> (tuple-length (Concat expr1 expr2)) 0))
      ((Get (Concat expr1 expr2) 0))
      :ruleset always-run)
;; next get
(rule ((= concat (Concat expr1 expr2))
       (= len (tuple-length concat))
       (= ith (Get concat i))
       (< (+ i 1) len)
       )
       ((Get (Concat expr1 expr2) (+ 1 i)))
       :ruleset always-run)

;; descend left
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (< i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr1 i)))
       :ruleset always-run)
;; descend right
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (>= i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr2 (- i len1))))
       :ruleset always-run)


(relation IsLeaf (Expr))
(rule ((Arg ty)) ((IsLeaf (Arg ty))) :ruleset always-run)
(rule ((Const val ty)) ((IsLeaf (Const val ty))) :ruleset always-run)
(rule ((Empty ty)) ((IsLeaf (Empty ty))) :ruleset always-run)

(rewrite (Concat (Concat a b) c)
         (Concat a (Concat b c))
         :ruleset always-run)

;;                tuple start len
(function SubTuple (Expr i64 i64) Expr :unextractable)
(rewrite (SubTuple expr x 1)
         (Single (Get expr x)) :ruleset always-run)
(rule ((= f (SubTuple expr a b)) (> b 1))
      ((union f 
         (Concat (Single (Get expr a))
                 (SubTuple expr (+ a 1) (- b 1))))) :ruleset always-run)