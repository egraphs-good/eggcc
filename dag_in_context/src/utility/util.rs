#[cfg(test)]
use crate::{egglog_test, interpreter::Value};

#[test]
fn test_list_util() -> crate::Result {
    use crate::ast::emptyt;
    use crate::schema::Assumption;
    let emptyt = emptyt();
    let ctx = Assumption::dummy();
    let build = format!(
        "
		(let list (Cons (Const (Int 0) {emptyt} {ctx})
                  (Cons (Const (Int 1) {emptyt} {ctx})
                  (Cons (Const (Int 2) {emptyt} {ctx})
                  (Cons (Const (Int 3) {emptyt} {ctx})
                  (Cons (Const (Int 4) {emptyt} {ctx}) (Nil)))))))
		(let expr (Switch (Const (Int 1) {emptyt} {ctx}) (Empty {emptyt} {ctx}) list))
	"
    );
    let check = format!(
        "
		(check (= (ListExpr-ith list 1) (Const (Int 1) {emptyt} {ctx})))
        (check (= (ListExpr-ith list 4) (Const (Int 4) {emptyt} {ctx})))
        (check (= (ListExpr-length list) 5))
	"
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn append_test() -> crate::Result {
    use crate::ast::emptyt;
    use crate::schema::Assumption;
    let emptyt = emptyt();
    let ctx = Assumption::dummy();
    let build = format!(
        "
        (let appended
            (Append
                (Cons (Const (Int 0) {emptyt} {ctx}) (Cons (Const (Int 1) {emptyt} {ctx}) (Nil)))
                (Const (Int 2) {emptyt} {ctx})))
    "
    );

    let check = format!("
        (check (
            =
            (Cons (Const (Int 0) {emptyt} {ctx}) (Cons (Const (Int 1) {emptyt} {ctx}) (Cons (Const (Int 2) {emptyt} {ctx}) (Nil))))
            appended
        ))
    ");
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_tuple_ith() -> crate::Result {
    use crate::ast::emptyt;
    use crate::schema::Assumption;
    let emptyt = emptyt();
    let ctx = Assumption::dummy();
    let build = format!(
        "
    (let tup (Concat
                  (Concat (Single (Const (Int 0) {emptyt} {ctx})) (Single (Const (Int 1) {emptyt} {ctx})))
                  (Concat (Single (Const (Int 2) {emptyt} {ctx})) (Single (Const (Int 3) {emptyt} {ctx})))))
    
    ;; Gets of tup and tup2 are only generated when tup and tup2 are in critical positions
    ;; like inputs to a DoWhile, 
    (let tup_type (TupleT (TCons (IntT) (TCons (IntT) (TCons (IntT) (TCons (IntT) (TNil)))))))
    (let body (Concat (Const (Bool true) tup_type (TmpCtx)) (Arg tup_type (TmpCtx))))
    (let loop (DoWhile tup body))
    (union (TmpCtx) (InLoop tup body))
    (delete (TmpCtx))
    
    ;; with print
    (let tup2 (Concat 
                (Concat 
                    (Single (Bop (Print) (Const (Int 0) {emptyt} {ctx}) (Arg (Base (StateT)) {ctx})))
                    (Concat (Single (Const (Int 1) {emptyt} {ctx})) 
                                (Single (Const (Int 2) {emptyt} {ctx}))))
                (Single (Const (Int 3) {emptyt} {ctx}))))
    (let tup2_type (TupleT (TCons (StateT) (TCons (IntT) (TCons (IntT) (TCons (IntT) (TNil)))))))
    (let body2 (Concat (Const (Bool true) tup2_type (TmpCtx)) (Arg tup2_type (TmpCtx))))
    (let loop2 (DoWhile tup2 body2))
    (union (TmpCtx) (InLoop tup2 body2))
    (delete (TmpCtx))
    "
    );

    let check = format!(
        "
    (check (= (Get tup 0) (Const (Int 0) {emptyt} {ctx})))
    (check (= (Get tup 1) (Const (Int 1) {emptyt} {ctx})))
    (check (= (Get tup 2) (Const (Int 2) {emptyt} {ctx})))
    (check (= (Get tup 3) (Const (Int 3) {emptyt} {ctx})))
    (check (= 4 (tuple-length tup)))
    (fail (check (Get tup 4)))

    (check (= (Get tup2 0) (Bop (Print) (Const (Int 0) {emptyt} {ctx}) (Arg (Base (StateT)) {ctx}))))
    (check (= (Get tup2 1) (Const (Int 1) {emptyt} {ctx})))
    (check (= (Get tup2 2) (Const (Int 2) {emptyt} {ctx})))
    (check (= (Get tup2 3) (Const (Int 3) {emptyt} {ctx})))
    (check (= 4 (tuple-length tup2)))
    (fail (check (Get tup2 4)))
    "
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}
