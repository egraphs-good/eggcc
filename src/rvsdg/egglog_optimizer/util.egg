;; ####################################
;; implementation of PassThroughArguments
;; Creates a vec of arguments
;; (vec-of (Arg 0) (Arg 1) ...) with length i
(function PassThroughArguments (i64) VecOperand :unextractable)



;; (how many arguments to generate, vector so far)
(function PassThroughArgumentsHelper (i64 VecOperandBase) VecOperand :unextractable)

(rewrite (PassThroughArguments i)
  (PassThroughArgumentsHelper i (vec-of))
  :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i rest))
       (< (vec-length rest) i))
      ((union lhs
        (PassThroughArgumentsHelper i
            (vec-push rest (Arg (vec-length rest))))))
      :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i rest))
       (= (vec-length rest) i))
      ((union lhs (VO rest)))
      :ruleset subst)

;; Project each argument out of a body
(function BodyToVecOperand (i64 Body) VecOperand :unextractable)
;; current index, body length, body, and vector so far
(function BodyToVecOperandHelper (i64 i64 Body VecOperandBase) VecOperand :unextractable)

(rewrite (BodyToVecOperand body-len body)
         (BodyToVecOperandHelper 0 body-len body (vec-of)) :ruleset subst)
(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (< index body-len))
  ((union helper
          (BodyToVecOperandHelper (+ index 1) body-len body
                                  (vec-push so-far 
                                   (Project index body)))))
  :ruleset subst)

(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (= index body-len))
  ((union helper (VO so-far)))
  :ruleset subst)


