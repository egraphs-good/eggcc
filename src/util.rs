use crate::{EggCCError, Optimizer};
use bril_rs::Program;
use std::fmt::Debug;
use std::{
    ffi::OsStr,
    fmt::{Display, Formatter},
    io,
    path::PathBuf,
    str::FromStr,
};

pub(crate) struct ListDisplay<'a, TS>(pub TS, pub &'a str);

impl<'a, TS> Display for ListDisplay<'a, TS>
where
    TS: Clone + IntoIterator,
    TS::Item: Display,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut did_something = false;
        for item in self.0.clone().into_iter() {
            if did_something {
                f.write_str(self.1)?;
            }
            Display::fmt(&item, f)?;
            did_something = true;
        }
        Ok(())
    }
}

/// Parse a string containing a bril program (in text format) into a Program.
///
/// This function is intended for use in tests and in ad-hoc debugging.
#[allow(unused)]
pub(crate) fn parse_from_string(input: &str) -> bril_rs::Program {
    use bril2json::parse_abstract_program_from_read;
    use bril_rs::load_program_from_read;
    let abs_program = parse_abstract_program_from_read(input.as_bytes(), true, false, None);
    let mut buf = Vec::new();
    serde_json::to_writer_pretty(&mut buf, &abs_program).unwrap();
    buf.push(b'\n');
    let json_str = String::from_utf8(buf).unwrap();
    load_program_from_read(json_str.as_bytes())
}

/// Write the visualizations to output files in the output directory.
/// If the directory does not exist, it creates it.
/// If the directory contains any files whose names conflict with the
/// output files, it replaces them.
///
/// Like other utilities related to `DebugVisualizations`, this method is
/// only intended for debugging eggcc.
pub fn visualize(test: TestProgram, output_dir: PathBuf) -> io::Result<()> {
    use std::fs::File;
    use std::io::Write;

    // make the directory if it doesn't exist
    if !output_dir.exists() {
        std::fs::create_dir_all(&output_dir)?;
    }

    let all_configs = Run::all_configurations_for(test);

    let results = all_configs.iter().map(|run| (run, run.run()));

    for (run, result) in results {
        let result = match result {
            Ok(res) => res,
            Err(err) => {
                eprintln!("Error running {}: {}", run.test_type, err);
                continue;
            }
        };
        // if there's an interpreted value do that as well
        if let Some(interpreted) = result.result_interpreted {
            let mut output_path = output_dir.clone();
            output_path.push(format!("{}-interp.txt", run.name()));
            let mut file = File::create(output_path)?;
            file.write_all(interpreted.as_bytes())?;
        }

        for visualization in result.visualizations {
            let mut output_path = output_dir.clone();

            assert!(output_path.is_dir());
            output_path.push(format!(
                "{}{}{}",
                run.name(),
                visualization.name,
                visualization.file_extension
            ));
            let mut file = File::create(output_path)?;
            file.write_all(visualization.result.as_bytes())?;
        }
    }

    Ok(())
}

/// Invokes some program with the given arguments, piping the given input to the program.
/// Returns an error if the program returns a non-zero exit code.
/// Code adapted from https://github.com/egraphs-good/egg/blob/e7845c5ae34267256b544c8e6b5bc36d91d096d2/src/dot.rs#L127
pub fn run_cmd_line<S1, S2, I>(program: S1, args: I, input: &str) -> std::io::Result<String>
where
    S1: AsRef<OsStr>,
    S2: AsRef<OsStr>,
    I: IntoIterator<Item = S2>,
{
    use std::io::Write;
    use std::process::{Command, Stdio};
    let mut child = Command::new(program)
        .args(args)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .spawn()?;

    let stdin = child.stdin.as_mut().expect("Failed to open stdin");
    write!(stdin, "{}", input)?;

    let output = child.wait_with_output()?;
    match output.status.code() {
        Some(0) => Ok(String::from_utf8(output.stdout).map_err(|e| {
            std::io::Error::new(std::io::ErrorKind::Other, format!("utf8 error: {}", e))
        })?),
        Some(e) => Err(std::io::Error::new(
            std::io::ErrorKind::Other,
            format!("program returned error code {}", e),
        )),
        None => Err(std::io::Error::new(
            std::io::ErrorKind::Other,
            "program was killed by a signal",
        )),
    }
}

/// Different ways to run eggcc- the default is RvsdgOptimize,
/// but these others are useful for testing and debugging.
#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub enum RunType {
    /// Do nothing to the input bril program besides parse it.
    /// Output the original program.
    Nothing,
    /// Convert the input bril program to a structured
    /// format.
    /// Output a human-readable debug version of this structured
    /// program, using while loops and if statements.
    StructuredConversion,
    /// Convert the input bril program to
    /// a structured format, then
    /// back to bril again.
    StructuredRoundTrip,
    /// Convert the input bril program to an RVSDG and output it as an SVG.
    RvsdgConversion,
    /// Convert to RVSDG and back to Bril again,
    /// outputting the bril program.
    RvsdgRoundTrip,
    /// Convert the original program to a CFG and output it as one SVG per function.
    ToCfg,
    /// Convert the original program to a CFG and back to Bril again.
    CfgRoundTrip,
    /// Removes unecessary direct
    /// jumps from the input program by
    /// converting it to a CFG, calling
    /// optimize_jumps, then converting it back to bril.
    OptimizeDirectJumps,
    /// Convert the original program to a RVSDG and then to a CFG, outputting one SVG per function.
    RvsdgToCfg,
    /// Convert the original program to a RVSDG, optimize it, then turn
    /// it back into a Bril program. This is the main way to run eggcc.
    RvsdgOptimize,
    /// Convert the original program to a RVSDG, optimize it, then output
    /// the optimized RVSDG as an SVG.
    OptimizedRvsdg,
    /// Convert the original program to a RVSDG, then output the egglog program
    /// that is used to optimize the RVSDG.
    RvsdgEgglogEncoding,
    /// Optimize the input as an RVSDG, and output the resulting
    /// egraph as an egraph-serialize crate json.
    Serialize,
}

impl Debug for RunType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self)
    }
}

impl FromStr for RunType {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "nothing" => Ok(RunType::Nothing),
            "structured" => Ok(RunType::StructuredConversion),
            "structured-roundtrip" => Ok(RunType::StructuredRoundTrip),
            "rvsdg" => Ok(RunType::RvsdgConversion),
            "rvsdg-roundtrip" => Ok(RunType::RvsdgRoundTrip),
            "to-cfg" => Ok(RunType::ToCfg),
            "cfg-roundtrip" => Ok(RunType::CfgRoundTrip),
            "optimize-direct-jumps" => Ok(RunType::OptimizeDirectJumps),
            "rvsdg-to-cfg" => Ok(RunType::RvsdgToCfg),
            "rvsdg-optimize" => Ok(RunType::RvsdgOptimize),
            "rvsdg-egglog-encoding" => Ok(RunType::RvsdgEgglogEncoding),
            "optimized-rvsdg" => Ok(RunType::OptimizedRvsdg),
            "serialize" => Ok(RunType::Serialize),
            _ => Err(format!("Unknown run type: {}", s)),
        }
    }
}

impl Display for RunType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            RunType::Nothing => write!(f, "nothing"),
            RunType::StructuredConversion => write!(f, "structured"),
            RunType::StructuredRoundTrip => write!(f, "structured-roundtrip"),
            RunType::RvsdgConversion => write!(f, "rvsdg"),
            RunType::RvsdgRoundTrip => write!(f, "rvsdg-roundtrip"),
            RunType::ToCfg => write!(f, "to-cfg"),
            RunType::CfgRoundTrip => write!(f, "cfg-roundtrip"),
            RunType::OptimizeDirectJumps => write!(f, "optimize-direct-jumps"),
            RunType::RvsdgToCfg => write!(f, "rvsdg-to-cfg"),
            RunType::RvsdgOptimize => write!(f, "rvsdg-optimize"),
            RunType::OptimizedRvsdg => write!(f, "optimized-rvsdg"),
            RunType::RvsdgEgglogEncoding => write!(f, "rvsdg-egglog-encoding"),
            RunType::Serialize => write!(f, "serialize"),
        }
    }
}

impl RunType {
    pub fn produces_bril(&self) -> bool {
        match self {
            RunType::Nothing => true,
            RunType::StructuredConversion => false,
            RunType::StructuredRoundTrip => true,
            RunType::RvsdgConversion => false,
            RunType::RvsdgRoundTrip => true,
            RunType::ToCfg => true,
            RunType::CfgRoundTrip => true,
            RunType::OptimizeDirectJumps => true,
            RunType::RvsdgToCfg => true,
            RunType::RvsdgOptimize => true,
            RunType::RvsdgEgglogEncoding => true,
            RunType::OptimizedRvsdg => false,
            RunType::Serialize => false,
        }
    }
}

#[derive(Clone)]
pub struct ProgWithArguments {
    program: Program,
    name: String,
    args: Vec<String>,
}

#[derive(Clone)]
pub enum TestProgram {
    Prog(ProgWithArguments),
    File(PathBuf),
}

impl TestProgram {
    pub fn read_program(self) -> ProgWithArguments {
        match self {
            TestProgram::Prog(prog) => prog,
            TestProgram::File(path) => {
                let program_read = std::fs::read_to_string(path.clone()).unwrap();
                let args = Optimizer::parse_bril_args(&program_read);
                let program = Optimizer::parse_bril(&program_read).unwrap();
                let name = path.file_stem().unwrap().to_str().unwrap().to_string();

                ProgWithArguments {
                    program,
                    name,
                    args,
                }
            }
        }
    }
}

#[derive(Clone)]
pub struct Run {
    pub prog_with_args: ProgWithArguments,
    pub test_type: RunType,
    // Also interpret the resulting program
    pub interp: bool,
    pub profile_out: Option<PathBuf>,
}

/// Some sort of visualization of the result, with a name
/// and a file extension.
/// For CFGs, the name is the name of the function and the vizalization
/// is a SVG.
#[derive(Clone)]
pub struct Visualization {
    pub result: String,
    pub file_extension: String,
    pub name: String,
}

#[derive(Clone)]
pub struct RunOutput {
    pub visualizations: Vec<Visualization>,
    // if the result was interpreted, the stdout of interpreting it
    pub result_interpreted: Option<String>,
    pub original_interpreted: Option<String>,
}

impl Run {
    pub fn all_configurations_for(test: TestProgram) -> Vec<Run> {
        let prog = test.read_program();
        let mut res = vec![];
        for test_type in [
            RunType::StructuredConversion,
            RunType::StructuredRoundTrip,
            RunType::RvsdgConversion,
            RunType::RvsdgRoundTrip,
            RunType::CfgRoundTrip,
            RunType::OptimizeDirectJumps,
            RunType::RvsdgOptimize,
            RunType::RvsdgToCfg,
            RunType::OptimizedRvsdg,
        ] {
            let default = Run {
                test_type,
                interp: false,
                prog_with_args: prog.clone(),
                profile_out: None,
            };
            res.push(default.clone());
            if test_type.produces_bril() {
                let interp = Run {
                    interp: true,
                    ..default
                };
                res.push(interp);
            }
        }
        res
    }

    // give a unique name for this run configuration
    pub fn name(&self) -> String {
        let mut name = format!("{}-{}", self.prog_with_args.name, self.test_type);
        if self.interp {
            name = format!("{}-interp", name);
        }
        name
    }

    pub fn run(&self) -> Result<RunOutput, EggCCError> {
        let original_interpreted = if self.interp {
            Some(Optimizer::interp(
                &self.prog_with_args.program,
                self.prog_with_args.args.clone(),
                None,
            ))
        } else {
            None
        };

        let (visualizations, bril_out) = match self.test_type {
            RunType::Nothing => (vec![], Some(self.prog_with_args.program.clone())),
            RunType::StructuredConversion => {
                let structured = Optimizer::program_to_structured(&self.prog_with_args.program)?;
                (
                    vec![Visualization {
                        result: structured.to_string(),
                        file_extension: ".txt".to_string(),
                        name: "".to_string(),
                    }],
                    None,
                )
            }
            RunType::StructuredRoundTrip => {
                let structured = Optimizer::program_to_structured(&self.prog_with_args.program)?;
                let bril = structured.to_program();
                (
                    vec![Visualization {
                        result: bril.to_string(),
                        file_extension: ".bril".to_string(),
                        name: "".to_string(),
                    }],
                    Some(bril),
                )
            }
            RunType::RvsdgConversion => {
                let rvsdg = Optimizer::program_to_rvsdg(&self.prog_with_args.program)?;
                let svg = rvsdg.to_svg();
                (
                    vec![Visualization {
                        result: svg,
                        file_extension: ".svg".to_string(),
                        name: "".to_string(),
                    }],
                    None,
                )
            }
            RunType::RvsdgRoundTrip => {
                let rvsdg = Optimizer::program_to_rvsdg(&self.prog_with_args.program)?;
                let cfg = rvsdg.to_cfg();
                let bril = cfg.to_bril();
                (
                    vec![Visualization {
                        result: bril.to_string(),
                        file_extension: ".bril".to_string(),
                        name: "".to_string(),
                    }],
                    Some(bril),
                )
            }
            RunType::RvsdgToCfg => {
                let rvsdg = Optimizer::program_to_rvsdg(&self.prog_with_args.program)?;
                let cfg = rvsdg.to_cfg();
                (cfg.visualizations(), None)
            }
            RunType::ToCfg => {
                let cfg = Optimizer::program_to_cfg(&self.prog_with_args.program);
                (cfg.visualizations(), None)
            }
            RunType::CfgRoundTrip => {
                let cfg = Optimizer::program_to_cfg(&self.prog_with_args.program);
                let bril = cfg.to_bril();
                (
                    vec![Visualization {
                        result: bril.to_string(),
                        file_extension: ".bril".to_string(),
                        name: "".to_string(),
                    }],
                    Some(bril),
                )
            }
            RunType::OptimizeDirectJumps => {
                let cfg = Optimizer::program_to_cfg(&self.prog_with_args.program);
                let optimized = cfg.optimize_jumps();
                let bril = optimized.to_bril();
                (
                    vec![Visualization {
                        result: bril.to_string(),
                        file_extension: ".bril".to_string(),
                        name: "".to_string(),
                    }],
                    Some(bril),
                )
            }
            RunType::RvsdgOptimize => {
                let optimized = Optimizer::rvsdg_optimize(&self.prog_with_args.program)?;
                (
                    vec![Visualization {
                        result: optimized.to_string(),
                        file_extension: ".bril".to_string(),
                        name: "".to_string(),
                    }],
                    Some(optimized),
                )
            }
            RunType::RvsdgEgglogEncoding => {
                let rvsdg = Optimizer::program_to_rvsdg(&self.prog_with_args.program)?;
                let (egglog_code, _) = rvsdg.build_egglog_code();
                (
                    vec![Visualization {
                        result: egglog_code,
                        file_extension: ".egglog".to_string(),
                        name: "".to_string(),
                    }],
                    None,
                )
            }
            RunType::OptimizedRvsdg => {
                let rvsdg = Optimizer::program_to_rvsdg(&self.prog_with_args.program)?;
                let optimized = rvsdg.optimize()?;
                let svg = optimized.to_svg();
                (
                    vec![Visualization {
                        result: svg,
                        file_extension: ".svg".to_string(),
                        name: "".to_string(),
                    }],
                    None,
                )
            }
            RunType::Serialize => {
                let rvsdg = Optimizer::program_to_rvsdg(&self.prog_with_args.program)?;
                let serialized = rvsdg.serialized_egraph()?;
                (
                    vec![Visualization {
                        result: serde_json::to_string_pretty(&serialized).unwrap(),
                        file_extension: ".json".to_string(),
                        name: "".to_string(),
                    }],
                    None,
                )
            }
        };

        let result_interpreted = match bril_out {
            Some(program) if self.interp => {
                assert!(self.test_type.produces_bril());
                Some(Optimizer::interp(
                    &program,
                    self.prog_with_args.args.clone(),
                    self.profile_out.clone(),
                ))
            }
            _ => None,
        };

        Ok(RunOutput {
            visualizations,
            result_interpreted,
            original_interpreted,
        })
    }
}

pub(crate) struct FreshNameGen {
    next: usize,
}

impl FreshNameGen {
    pub(crate) fn new() -> Self {
        Self { next: 0 }
    }

    pub(crate) fn fresh(&mut self) -> String {
        let name = format!("v{}", self.next);
        self.next += 1;
        name
    }

    pub(crate) fn fresh_usize(&mut self) -> usize {
        let name = self.next;
        self.next += 1;
        name
    }
}
