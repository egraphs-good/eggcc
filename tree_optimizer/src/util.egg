(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((All id order top)) ((union (ListExpr-suffix top 0) top)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
    (Cons a (Append b e))
    :ruleset always-run)
(rewrite (Append (Nil) e)
    (Cons e (Nil))
    :ruleset always-run)

(function Expr-size (Expr) i64 :merge (min old new))
(function ListExpr-size (ListExpr) i64 :merge (min old new))
