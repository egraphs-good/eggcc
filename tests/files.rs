use std::collections::HashSet;

use eggcc::util::{InterpMode, Run, RunType, TestProgram};
use insta::assert_snapshot;
use libtest_mimic::Trial;

/// Generate tests for all configurations of a given file
/// If `just_brilift` is true, only generate tests that
/// run the full pipeline with brilift
fn generate_tests(glob: &str, benchmark_mode: bool) -> Vec<Trial> {
    let mut trials = vec![];

    let mut mk_trial = |run: Run, snapshot: bool| {
        let snapshot_configurations: HashSet<RunType> = [RunType::Optimize].into_iter().collect();

        trials.push(Trial::test(run.name(), move || {
            let result = match run.run() {
                Err(error) => {
                    panic!("{}", error);
                }
                Ok(res) => res,
            };
            if run.test_type == RunType::CompileBrilift || run.test_type == RunType::CompileBrilLLVM
            {
                let executable = run
                    .output_path
                    .clone()
                    .unwrap_or_else(|| format!("/tmp/{}", run.name()));
                std::process::Command::new("rm")
                    .args(vec![executable.clone(), executable + "-args"])
                    .status()
                    .unwrap();
            }

            if run.interp.should_interp()
                && result.original_interpreted.as_ref().unwrap()
                    != result.result_interpreted.as_ref().unwrap()
            {
                panic!(
                    "Interpreted result does not match expected:\nExpected: {}\nGot: {}",
                    result.original_interpreted.unwrap(),
                    result.result_interpreted.unwrap()
                );
            }
            // only assert a snapshot if we are in the "small" folder
            if snapshot && snapshot_configurations.contains(&run.test_type) {
                for visualization in result.visualizations {
                    assert_snapshot!(run.name() + &visualization.name, visualization.result);
                }
            }
            Ok(())
        }))
    };

    for entry in glob::glob(glob).unwrap() {
        let f = entry.unwrap();

        let snapshot = f.to_str().unwrap().contains("small");

        // in benchmark mode, run fewer tests and use binaries to interpret
        let configurations = if benchmark_mode {
            // use InterpFast, which runs brilift
            // with optimization but without egglog optimization
            vec![
                Run::compile_brilift_config(
                    TestProgram::BrilFile(f.clone()),
                    true,
                    true,
                    InterpMode::InterpFast,
                ),
                Run::compile_llvm_config(
                    TestProgram::BrilFile(f.clone()),
                    true,
                    true,
                    InterpMode::InterpFast,
                ),
                Run::compile_llvm_config(
                    TestProgram::BrilFile(f),
                    false,
                    true,
                    InterpMode::InterpFast,
                ),
            ]
        } else {
            Run::all_configurations_for(TestProgram::BrilFile(f))
        };

        for run in configurations {
            mk_trial(run, snapshot);
        }
    }

    trials
}

fn main() {
    let args = libtest_mimic::Arguments::from_args();
    let mut tests = generate_tests("tests/passing/**/*.bril", false);
    // also generate tests for benchmarks
    tests.extend(generate_tests("benchmarks/passing/**/*.bril", true));

    libtest_mimic::run(&args, tests).exit();
}
