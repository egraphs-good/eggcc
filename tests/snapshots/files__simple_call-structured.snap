---
source: tests/files.rs
expression: visualization.result
---
inc {
one: int = const 1;
res: int = add one x;
print res;
return
}

main {
a: int = const 0;
call @inc a;
return
}
